### PR TITLE
gh-124218: Use per-thread refcounts for code objects

### PR DESCRIPTION
**Pull request by colesbury**
Use per-thread refcounting for the reference from function objects to their corresponding code object. This can be a source of contention when frequently creating nested functions. Deferred refcounting alone isn't a great fit here because these references are on the heap and may be modified by other libraries.

We will still need to address the globals and builtins dictionaries in a similar way.

* Issue: gh-124218

### CODE REVIEW

**Comment by markshannon** [comment]
This seems like we are adding more code, rather than reusing the existing code for per-thread refcounts already in existence for classes.
Can we factor out this code?

**Comment by colesbury** [comment]
It is factored out and reusing existing code. That's why `_Py_INCREF_CODE` and `_Py_DECREF_CODE` are only one line each.

What concretely would you like changed?

**Comment by markshannon** [comment]
Quite a lot of code is being added to pycore_object.h, but I don't see code being removed from typeobject.c or wherever the code for per-thread reference counts for classes is.
I would have though that classes could use the new code as well.

**Comment by colesbury** [comment]
What would you like to see changed? `pycore_object.h` is +22 lines. That doesn't seem like quite a lot to me. That is:

* Add two simple functions with one line bodies (`_Py_INCREF_CODE` and `_Py_DECREF_CODE`): +12
* Split off the generic `_Py_THREAD_INCREF_OBJECT` from `_Py_INCREF_TYPE`: +5
* Split off the generic `_Py_THREAD_DECREF_OBJECT` from `_Py_DECREF_TYPE`: +5

File: Python/uniqueid.c [hunk 1]
```diff
@@ -179,20 +203,15 @@ _PyObject_FinalizeUniqueIdPool(PyInterpreterState *interp)
         pool->freelist = next;
     }
 
-    // Now everything non-NULL is a type. Set the type's id to -1 in case it
-    // outlives the interpreter.
+    // Now everything non-NULL is a object. Clear their unique ids case the
```
Comment (kumaraditya303) [inline Python/uniqueid.c]:
```suggestion
    // Now everything non-NULL is a object. Clear their unique ids as the
```

File: Include/internal/pycore_object.h [hunk 2]
```diff
@@ -361,25 +394,14 @@ _Py_DECREF_TYPE(PyTypeObject *type)
...
-    }
-    else {
-        // Directly decref the type if the type id is not assigned or if
-        // per-thread refcounting has been disabled on this type.
-        Py_DECREF(type);
-    }
+static inline void
+_Py_DECREF_CODE(PyCodeObject *co)
+{
+    _Py_THREAD_DECREF_OBJECT((PyObject *)co, co->_co_unique_id);
```
Comment (kumaraditya303) [inline Include/internal/pycore_object.h:404]: assert that co is a code object.

File: Include/internal/pycore_object.h [hunk 3]
```diff
@@ -328,29 +352,38 @@ _Py_INCREF_TYPE(PyTypeObject *type)
...
+    _Py_THREAD_INCREF_OBJECT((PyObject *)type, ((PyHeapTypeObject *)type)->unique_id);
+#if defined(__GNUC__) && __GNUC__ >= 11
+#  pragma GCC diagnostic pop
+#endif
+}
 
+static inline void
+_Py_INCREF_CODE(PyCodeObject *co)
+{
+    _Py_THREAD_INCREF_OBJECT((PyObject *)co, co->_co_unique_id);
```
Comment (kumaraditya303) [inline Include/internal/pycore_object.h:364]: Same add assert to check it is codeobject

File: Include/internal/pycore_object.h [hunk 4]
```diff
@@ -311,7 +311,31 @@ extern bool _PyRefchain_IsTraced(PyInterpreterState *interp, PyObject *obj);
...
+    // the "else".
+    if ((size_t)unique_id < (size_t)tstate->refcounts.size) {
+#  ifdef Py_REF_DEBUG
+        _Py_INCREF_IncRefTotal();
+#  endif
+        _Py_INCREF_STAT_INC();
+        tstate->refcounts.values[unique_id]++;
+    }
+    else {
+        // The slow path resizes the thread-local refcount array if necessary.
```
Comment (mpage) [inline Include/internal/pycore_object.h]:
```suggestion
        // The slow path resizes the per-thread refcount array if necessary.
```

**Comment by colesbury** [comment]
!buildbot AMD64 Fedora Stable Refleaks

File: Include/internal/pycore_object.h (hunk 2 above)
Comment (colesbury) [reply Include/internal/pycore_object.h:404]: `co` is statically a `PyCodeObject *`. We don't typically dynamically check the type when we have the static type (e.g., in [`PyCode_Addr2Line`](https://github.com/python/cpython/blob/aac89b54c5ee03c4d64fbdfbb6ea3001e26aa83a/Objects/codeobject.c#L941-L942))

File: Include/internal/pycore_object.h [hunk 5]
```diff
@@ -328,29 +352,38 @@ _Py_INCREF_TYPE(PyTypeObject *type)
...
-        tstate->refcounts.values[ht->unique_id]++;
+        _Py_DECREF_STAT_INC();
+        tstate->refcounts.values[unique_id]--;
     }
     else {
-        // The slow path resizes the thread-local refcount array if necessary.
-        // It handles the unique_id=-1 case to keep the inlinable function smaller.
-        _PyType_IncrefSlow(ht);
+        // Directly decref the type if the type id is not assigned or if
+        // per-thread refcounting has been disabled on this type.
```
Comment (mpage) [inline Include/internal/pycore_object.h]:
```suggestion
        // Directly decref the object if the id is not assigned or if
        // per-thread refcounting has been disabled on this object.
```

File: Python/uniqueid.c [hunk 6]
```diff
@@ -98,36 +98,60 @@ _PyObject_AssignUniqueId(PyObject *obj)
...
+}
+
 void
-_PyType_IncrefSlow(PyHeapTypeObject *type)
+_PyObject_ThreadIncrefSlow(PyObject *obj, Py_ssize_t unique_id)
 {
     _PyThreadStateImpl *tstate = (_PyThreadStateImpl *)_PyThreadState_GET();
-    if (type->unique_id < 0 || resize_local_refcounts(tstate) < 0) {
+    if (unique_id < 0 || resize_local_refcounts(tstate) < 0) {
         // just incref the type directly.
```
Comment (mpage) [inline Python/uniqueid.c]:
```suggestion
        // just incref the object directly.
```

**Review by mpage — APPROVED** [review]
LGTM - just some comment fixups inline.

File: Include/internal/pycore_object.h (hunk 2 above)
Comment (kumaraditya303) [reply Include/internal/pycore_object.h:404]: I meant it to be protect against incorrect casts, anyways it's minor.

**Review by kumaraditya303 — APPROVED** [review]

**Pull request merged**